scipy/spatial/transform/rotation.py

```diff
@@ -8,7 +8,102 @@
 AXIS_TO_IND = {'x': 0, 'y': 1, 'z': 2}
 
 
-def _make_elementary_quat(axis, angles):
+def elementary_basis_vector(axis):
+    b = np.zeros(3)
+    b[AXIS_TO_IND[axis]] = 1
+    return b
+
+
+def compute_euler_from_dcm(dcm, seq, extrinsic=False):
+    # The algorithm assumes intrinsic frame transformations. For representation
+    # the paper uses transformation matrices, which are transpose of the
+    # direction cosine matrices used by our Rotation class.
+    # Adapt the algorithm for our case by
+    # 1. Transposing our dcm representation
+    # 2. Reversing both axis sequence and angles for extrinsic rotations
+
+    if extrinsic:
+        seq = seq[::-1]
+
+    if dcm.ndim == 2:
+        dcm = dcm[None, :, :]
+    num_rotations = dcm.shape[0]
+
+    # Step 0
+    # Algorithms dcms are transpose of ours
+    dcm = np.transpose(dcm, (0, 2, 1))
+    # Algorithm assumes axes as column vectors, here we use 1D vectors
+    n1 = elementary_basis_vector(seq[0])
+    n2 = elementary_basis_vector(seq[1])
+    n3 = elementary_basis_vector(seq[2])
+
+    # Step 2
+    sl = np.dot(np.cross(n1, n2), n3)
+    cl = np.dot(n1, n3)
+
+    norm = sl**2 + cl**2
+    sl /= norm
+    cl /= norm
+
+    lamb = np.arctan2(sl, cl)
+    c = np.empty((3, 3))
+    c[0] = n2
+    c[1] = np.cross(n1, n2)
+    c[2] = n1
+
+    # Step 3
+    rt = np.array([
+        [1, 0, 0],
+        [0, cl, -sl],
+        [0, sl, cl]
+    ])
+    rtc = rt.dot(c)
+    ct = c.T
+    o = np.empty_like(dcm)
+    res = np.einsum('ij,...jk->...ik', rtc, dcm)
+    o = np.einsum('...ij,jk->...ik', res, ct)
+
+    # Step 4
+    angle2 = lamb + np.arccos(o[:, 2, 2])
+
+    # Steps 5, 6
+    eps = np.finfo(float).resolution  # ~1e-15
+    safe1 = (np.abs(angle2 - lamb) >= eps)
+    safe2 = (np.abs(angle2 - lamb - np.pi) >= eps)
+
+    angle1 = np.empty(num_rotations)
+    angle3 = np.empty(num_rotations)
+
+    # 5b
+    safe_mask = np.logical_and(safe1, safe2)
+    angle1[safe_mask] = np.arctan2(o[safe_mask, 2, 0], -o[safe_mask, 2, 1])
+    angle3[safe_mask] = np.arctan2(o[safe_mask, 0, 2], o[safe_mask, 1, 2])
+
+    # 6a
+    angle3[~safe_mask] = 0
+    # 6b
+    angle1[~safe1] = np.arctan2(o[~safe1, 0, 1] - o[~safe1, 1, 0],
+                                o[~safe1, 0, 0] + o[~safe1, 1, 1])
+    # 6c
+    angle1[~safe2] = np.arctan2(o[~safe2, 0, 1] + o[~safe2, 1, 0],
+                                o[~safe2, 0, 0] - o[~safe2, 1, 1])
+
+    # Step 7
+    # python modulo operator works correctly for negative numbers
+    adjust_mask = np.logical_or(angle2 < 0, angle2 > np.pi)
```

Comment:
OK, I experimented with this adjustment a little bit and I propose the following convention:

1. Second angle is from -90 to 90 degrees.
2. First and third angles are from -180 to 180 degrees. Could always achieve that, my point is not to use [0, 360], you need to modify your code a little bit.

I think this is a decent convention for common 3-1-3 and 3-2-1 (yaw, pitch, roll) sequence. At least for me it seems more natural than restricting the second angle from 0 to 180.

Reply:
Think it through or experiment --- is it possible that the second angle won't be in [-90, 90] when we apply the correction.

Reply:
I was just experimenting with this and I have a question. Is is possible to represent a single rotation about, say y-axis, with angle 135 degrees, as a rotation about different axes?
If not, then `Rotation.as_euler('zyx')` cannot have angle2 in range `[-90, 90]`.

Further, for axes sequence `zxz`, parameter `lambda` is `0`. This means that adjusting `angle2` to be within range `[-90, 90]` is not possible. For example, I'm trying to do `Rotation.from_euler('zxz', [25, 135, 35], degrees=True).as_euler('zxz', degrees=True)`. Since `lambda = 0` for this, `angle2` can only be `135` or `-135` degrees. I think this is why the paper chose the range of restriction to be `[0, 180]`. For now I'm restricting range of `angle2` to be `[0, 180]` as per the paper.

Reply:
Another update. So I tried this test case: `R.from_euler('zxy', [30, 100, 60], degrees=True).as_euler('zxy', degrees=True)`. The angles returned by the algorithm are `[-150.,   80., -120.]`, without any adjustments. I think we need to rethink our approach of checking for equality of angles. @nmayorov what do you suggest?

Reply:
Good investigation. I can't dig into it at the moment (will do tonight), I suggest you to consider which `phi` angles we get for different axis sequences. I will make a guess/suggestion that for `zyx` (and alike) axes we should constraint the second angle to [-90, 90] so it will behave conventionally for roll, pitch and yaw. And for 'zxz' we should constraint it into [0, 180] --- because we can't achieve [-90, 90] and it should also make sense. Please investigate it farther.